pkg/service/types.go

```diff
@@ -70,3 +72,36 @@ type WeightedCluster struct {
 	ClusterName ClusterName `json:"cluster_name:omitempty"`
 	Weight      int         `json:"weight:omitempty"`
 }
+
+// Provider is an interface to be implemented by components abstracting Kubernetes, and other compute/cluster providers
+type Provider interface {
+	// GetServicesForServiceAccount returns a list of services corresponding to a service account
+	GetServicesForServiceAccount(identity.K8sServiceAccount) ([]MeshService, error)
+
+	// ListServices returns a list of services that are part of monitored namespaces
+	ListServices() ([]MeshService, error)
+
+	// ListServiceIdentitiesForService returns service identities for given service
+	ListServiceIdentitiesForService(svc MeshService) ([]identity.ServiceIdentity, error)
+
+	// GetPortToProtocolMappingForService returns a mapping of the service's ports to their corresponding application protocol,
+	// where the ports returned are the ones used by downstream clients in their requests. This can be different from the ports
+	// actually exposed by the application binary, ie. 'spec.ports[].port' instead of 'spec.ports[].targetPort' for a Kubernetes service.
+	GetPortToProtocolMappingForService(svc MeshService) (map[uint32]string, error)
+
+	// GetTargetPortToProtocolMappingForService returns a mapping of the service's ports to their corresponding application protocol.
+	// The ports returned are the actual ports on which the application exposes the service derived from the service's endpoints,
+	// ie. 'spec.ports[].targetPort' instead of 'spec.ports[].port' for a Kubernetes service.
+	// The function ensures the port:protocol mapping is the same across different endpoint providers for the service, and returns
+	// an error otherwise.
+	GetTargetPortToProtocolMappingForService(svc MeshService) (map[uint32]string, error)
+
+	// GetHostnamesForService returns a list of hostnames over which the service can be accessed within the local cluster.
+	GetHostnamesForService(svc MeshService, locality Locality) ([]string, error)
+
+	// GetServicesByLabels gets Kubernetes services whose selectors match the given labels
+	GetServicesByLabels(podLabels map[string]string, namespace string) ([]MeshService, error)
+
+	// GetID returns the unique identifier of the EndpointsProvider.
+	GetID() string
```

Comment:
Seems like we don't need this at this moment. Can be removed.

Reply:
FYI it will be used for logging, similar to the endpoint provider's usage

Reply:
Comment needs to be updated, refers to `EndpointsProvider`.